### ISSUE TITLE
Showing accessibility frame while scrolling regressed scrolling on older devices when semantics are enabled

### ISSUE THREAD

**Issue by ds84182** [flutter/flutter#12138]
Commit: ecb5692

Use any list with text. The more text in the list, the worse the performance.

Flutter Gallery's List example and Grid example can be used to test performance.

When disabling text on the Grid example I get full 60fps. When enabling text on the Grid example I get at least 30fps. The engine and rasterizer do not report the performance drop, so it might be Flutter missing vsync.

**Comment by eseidelGoogle** [comment]
@goderbauer

**Comment by goderbauer** [comment]
It also regressed another benchmark according to @Hixie:

"flutter_gallery__transition_perf_with_semantics delta_missed_frame_build_budget_count regressed after #12039."

**Comment by goderbauer** [comment]
@Hixie After looking at the benchmark I am not surprised that it regressed. The benchmark includes scrolling while semantics are turned on. Before the change, we were cutting off all semantics during scrolling. With that change, we are now computing the semantics tree during scrolling (to display the accessibility focus indicator at the right position), which is obviously more expansive. So, I'd say it's an expected regression, which can (short of removing the feature) only be solved by making a11y tree generation more efficient in general (which is not that trivial). We should probably just re-base that benchmark for now.

@ds84182 What hardware are you running on? Do you have any accessibility features enabled? Our benchmarks (and my testing on device) shows no difference in performance when accessibility features are disabled with and without the patch. However, if you have accessibility features enabled, there will be a slight performance hit (see comment above).

**Comment by ds84182** [comment]
I am running on a Moto Z Play. I do have an app running that uses Accessibility features (Sesame Shortcuts).

Native Android applications don't experience any lag with accessibility enabled so I wouldn't expect a Flutter application to experience lag.

**Comment by goderbauer** [comment]
@ds84182 I cannot reproduce this on my Moto G4 (see screenshots below). Any chance you can upload a video comparing the scrolling performance you experience before and after the patch?

Scrolling performance **with** the patch in question (ecb5692):
![withpatch](https://user-images.githubusercontent.com/1227763/30560851-04a4d556-9c6e-11e7-9d66-149890d3ecee.png)

Scrolling performance **without** the patch:
![withoutpatch](https://user-images.githubusercontent.com/1227763/30560871-11dec11e-9c6e-11e7-862c-2b9c52bd74cb.png)

**Comment by Hixie** [comment]
Those tests were in debug builds, which probably are affecting your results. You should check the results in profile or release builds.

**Comment by goderbauer** [comment]
The absolute numbers are different in profile mode, but there is still no difference between having the patch and not having the patch:

Scrolling performance **with** the patch applied in profile mode:
![withpatch-release](https://user-images.githubusercontent.com/1227763/30563298-4d798b84-9c76-11e7-8c6c-b79b9eb40527.png)

Scrolling performance **without** the patch in profile mode:
![withoutpatch-release](https://user-images.githubusercontent.com/1227763/30563313-582065da-9c76-11e7-993e-de87e2837527.png)

**Comment by ds84182** [comment]
I'll have a video shortly, I'm busy with something right now.

The performance drop doesn't seem to happen in the rasterizer and engine since (or at least I believe) it happens outside of both of them in the AccessibilityNodeProvider. I did compare the performance overlay with before and after and I didn't observe any changes in the performance overlay.

I can also get a systrace of it, if needed.

**Comment by ds84182** [comment]
Here's a video at ecb5692:

Google Drive's preview might not be 60FPS, so you should download it.

https://drive.google.com/file/d/0BzkaO0atYpl3YWZBRFY1UmQ0eG8/view?usp=sharing

Here's a video for the commit before ecb5692:

Same as above, the preview might not be 60FPS.

https://drive.google.com/file/d/0BzkaO0atYpl3OGY3cTVlbnQ4YkU/view?usp=sharing

**Comment by goderbauer** [comment]
Thanks @ds84182 for the video. I was able to reproduce the performance regression on my G4 and it also shows up in the trace: 

[dart-timeline-2017-8-20.json.zip](https://github.com/flutter/flutter/files/1318739/dart-timeline-2017-8-20.json.zip)

We will have to investigate how we can improve the semantics algorithm to make updating the semantics tree more efficient during scrolling.

**Comment by rom1dep** [comment]
Hi there, I'm new to flutter, but was disappointed by the performances of the recent builds on my Xperia [E5803](https://en.wikipedia.org/wiki/Sony_Xperia_Z5_Compact) (Adreno 430, so a fairly beefy and not actually that "old" device).

I bisected, which pointed me to [ecb5692](https://github.com/flutter/flutter/commit/ecb56927deb0de807a6fd0619203004ed0e8e772) and to this issue.
This change pretty much makes every demo app almost unusable on my device, so I'll be happy to help with some testing if anything moves here.

**Comment by goderbauer** [comment]
I am going to close this one for now as there is not much we can do about it. We need the a11y support.
